### PR TITLE
[8.0] add account_invoice_product_supplier_price_update

### PR DESCRIPTION
**Pull request by chafique-delli**
This module allows in the supplier invoice, to automatically update all products whose unit price on the line is different from the supplier price.

*  **account_invoice_supplierinfo_update** : provides main features to update supplierinfo from invoice line information
* **account_invoice_supplierinfo_update_discount** : glue module to update discount field, if 'product_supplierinfo_discount' is installed.
* **account_invoice_supplierinfo_update_variant** : glue module to update variant supplierinfo, if 'product_variant_supplierinfo' is installed.
*  **account_invoice_supplierinfo_update_on_validate** : extra module to force user to update supplierinfo data when user validates the invoice.

### CODE REVIEW

File: account_invoice_product_supplier_price_update/README.rst [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/142/8.0
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/account-invoicing/issues>`_. In case of trouble, please
+check there if your issue has already been reported. If you spotted it first,
+help us smashing it by providing a detailed and welcomed `feedback
```
Comment (elicoidal) [inline account_invoice_product_supplier_price_update/README.rst]: You can simplify this part with the newest template

File: account_invoice_product_supplier_price_update/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
+# © 2016 Chafique DELLI @ Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    'name': 'Account Invoice Product Supplier Price Update',
+    'summary': 'From the supplier invoice, to automatic update all products '
```
Comment (elicoidal) [inline account_invoice_product_supplier_price_update/__openerp__.py]: In the supplier invoice, automatically update all products 
whose unit price on the line is different from the supplier price

File: account_invoice_product_supplier_price_update/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+               'whose unit price on the line is different from '
+               'the supplier price',
+    'version': '8.0.1.0.0',
+    'category': 'Accounting & Finance',
+    'website': 'http://akretion.com',
+    'author': 'Akretion, Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'application': False,
+    'installable': True,
```
Comment (elicoidal) [inline account_invoice_product_supplier_price_update/__openerp__.py]: probably not necessary as this is the default one

**Comment by chafique-delli** [comment]
@elicoidal , thank you for the code review, I fixed this.

File: account_invoice_product_supplier_price_update/README.rst [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+======================================================
+Update Supplier Price of product from Supplier Invoice
+======================================================
+
+This module allow in the supplier invoice, automatically update all products
+whose unit price on the line is different from the supplier price.
+It created a new supplier price if there is not or it updates the first.
```
Comment (eLBati) [inline account_invoice_product_supplier_price_update/README.rst]: `created` --> `creates`

**Comment by chafique-delli** [comment]
@eLBati , thank you for the code review, I fixed this.

File: account_invoice_product_supplier_price_update/README.rst [hunk 5]
```diff
@@ -0,0 +1,46 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+======================================================
+Update Supplier Price of product from Supplier Invoice
+======================================================
+
+This module allow in the supplier invoice, automatically update all products
```
Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/README.rst]: s/allow/allows

Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/README.rst]: s/, automatically/, to automatically

File: account_invoice_product_supplier_price_update/__openerp__.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    'version': '8.0.1.0.0',
+    'category': 'Accounting & Finance',
+    'website': 'http://akretion.com',
+    'author': 'Akretion, Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'installable': True,
+    'depends': [
+        'account',
+        'purchase',
+        'product_variant_supplierinfo',
```
Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/__openerp__.py]: This module and the following are not real dependencies. Just conveniences, but the module will work the same, except for that features.

File: account_invoice_product_supplier_price_update/__openerp__.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+               'the supplier price',
+    'version': '8.0.1.0.0',
+    'category': 'Accounting & Finance',
+    'website': 'http://akretion.com',
+    'author': 'Akretion, Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'installable': True,
+    'depends': [
+        'account',
+        'purchase',
```
Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/__openerp__.py]: purchase is not needed for having product.supplierinfo. They are in the product module.

File: account_invoice_product_supplier_price_update/__openerp__.py (hunk 6 above)
Comment (chafique-delli) [inline account_invoice_product_supplier_price_update/__openerp__.py]: @pedrobaeza ,  thank you for the code review.
Indeed, the dependencies on the modules 'purchase' and 'product_variant_supplierinfo' are not necessary.

Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/__openerp__.py]: And you don't need also `product_supplierinfo_tree_price_info`

Comment (chafique-delli) [inline account_invoice_product_supplier_price_update/__openerp__.py]: @pedrobaeza ,in my code, I use the field 'unit_price note' of 'product.supplierinfo' which comes from the module 'product_supplierinfo_tree_price_info'.

Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/__openerp__.py]: But that module is only for displaying information in an human readable form, not for using for another purposes. I would say to not using anything from this module.

**Comment by chafique-delli** [comment]
Travis test is red because the test of  'account_invoice_merge_purchase' module failed.

File: account_invoice_product_supplier_price_update/account_invoice.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    @api.model
+    def _check_product_supplierprice(self):
+        lines = []
+        for line in self.invoice_line:
+            current_price_unit = [(6, 0, [])]
+            suppinfo = self.env['product.supplierinfo'].search([
+                ('product_tmpl_id', '=', line.product_id.product_tmpl_id.id),
```
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/account_invoice.py]: you must depend on this module to get the right supplier https://github.com/OCA/product-variant/pull/32

**Comment by sebastienbeau** [comment]
@chafique-delli can you fis the way to get the right supplier and add a test with
- a specific supplier (a supplier info attach to one variant)
- a generic supplier (a supplier info attach to the template => all variant)

File: account_invoice_product_supplier_price_update/account_invoice.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    @api.model
+    def _check_product_supplierprice(self):
+        lines = []
+        for line in self.invoice_line:
+            current_price_unit = [(6, 0, [])]
```
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/account_invoice.py]: The code can be more simple here.
- first the current_price_unit do not need the 6, 0 as you always use the [0][2] after so you can remove it.
- then you do not need all of the variable suppinfo...

`
something like that:

```
def _check_product_supplierprice(self):                         
lines = []                                                             
for line in self.invoice_line:                                         
    current_price_unit = []
    supplier = self.partner_id.commercial_partner_id or self.partner_id
suppinfo = False
    current_price_uni = []
    for seller in line.product_id.seller_ids:                          
        if supplier == seller.name: 
            suppinfo = seller
            current_price_unit = [pricelist.price for p in suppinfo.pricelist_ids]
            break                                                      
    if line.price_unit not in current_price_unit:                                               
        lines.append((0, 0, {                                          
            'name': line.name,                                         
            'current_price_unit': suppinfo.unit_price_note,            
            'new_price_unit': line.price_unit,                         
            'suppinfo_id': suppinfo.id,                                
            'to_variant': True,                                        
            'supplier_id': supplier_id.id,                             
            'product_id': line.product_id.id,                          
            'product_tmpl_id': line.product_id.product_tmpl_id.id,     
        }))                                                            
return lines                                                           
```

`

File: account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+            else:
+                vals = {
+                    'suppinfo_id': line.suppinfo_id.id,
+                    'min_quantity': 0.0,
+                    'price': line.new_price_unit,
+                }
+                self.env['pricelist.partnerinfo'].create(vals)
+
+
+class SupplierinvoiceUpdateProductSupplierpriceLine(models.TransientModel):
```
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py]: can you found a shorter class name?

File: account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+                }
+                if line.to_variant:
+                    vals.update({
+                        'product_id': line.product_id.id,
+                    })
+                else:
+                    vals.update({
+                        'product_tmpl_id': line.product_id.product_tmpl_id.id,
+                    })
+                line.suppinfo_id = self.env['product.supplierinfo'].create(
```
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py]: can you avoid the repetition of self.env['product.supplierinfo'] by creating a variable called supinfo_obj ?

File: account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+# © 2016 Chafique DELLI @ Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import models, api, fields
+import openerp.addons.decimal_precision as dp
+
+
+class SupplierinvoiceUpdateProductSupplierprice(models.TransientModel):
```
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py]: can you find a shorter class name?

**Review by sebastienbeau — CHANGES_REQUESTED** [review]
Some code simplification should be done

File: account_invoice_product_supplier_price_update/tests/test_supplier_invoice_update_product_supplier_price.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+# © 2016 Chafique DELLI @ Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp.tests.common import TransactionCase
+
+
+class TestSupplierInvoiceUpdateProductSupplierPrice(TransactionCase):
```
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/tests/test_supplier_invoice_update_product_supplier_price.py]: We should split the test in 3 as there is 3 test on 3 invoice

**Review by sebastienbeau — COMMENTED** [review]
split test

File: account_invoice_product_supplier_price_update/tests/test_update_supplier_price.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+            pricelist_partnerinfos = self.env['pricelist.partnerinfo'].search([
+                ('suppinfo_id', '=', line.suppinfo_id.id),
+                ('min_quantity', '=', 0.0),
+                ('price', '=', line.new_price_unit)
+            ])
+            self.assertNotEquals(pricelist_partnerinfos, False)
+
+        # supplier invoice with pricelist supplierinfo to update and
+        # product supplierinfo is on product_product
+        invoice_7 = self.env.ref(
```
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/tests/test_update_supplier_price.py]: Can you split the test ? one per invoice

**Comment by chafique-delli** [comment]
@gurneyalex, Is you can watch why the runbot test shows me this error:
except_orm: ('Error', u"You try to install module 'account_invoice_product_supplier_price_update' that depends on module 'product_variant_supplierinfo'.\nBut the latter module is not available in your system.")
While the file 'oca_dependencies.txt' contains' product-variant' repository which contains  product_variant_supplierinfo' module.
Thank you.

**Comment by gurneyalex** [comment]
@chafique-delli runbot should be fixed now.

**Comment by chafique-delli** [comment]
@gurneyalex , thank you for the fix on runbot, but is you can watch why the travis test shows an error on installing the  account_invoice_merge_purchase module whereas,  a priori my  account_invoice_product_supplier_price_update module is not installed yet.

**Comment by gurneyalex** [comment]
@chafique-delli this is not your fault. The 8.0 branch build is red on Travis (https://travis-ci.org/OCA/account-invoicing/jobs/166920774)  -> #185

File: account_invoice_product_supplier_price_update/wizard/update_supplier_price.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    wizard_line_ids = fields.One2many(
+        'update.supplierprice.line',
+        'wizard_id',
+        string='Wizard lines')
+
+    @api.multi
+    def update_product_supplierprice(self):
+        invoice = self.env['account.invoice'].browse(
+            self._context['active_id'])
+        invoice.signal_workflow('invoice_open')
```
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/wizard/update_supplier_price.py]: Why do you launch this workflow ?

File: account_invoice_product_supplier_price_update/wizard/update_supplier_price.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    _name = 'update.supplierprice'
+
+    wizard_line_ids = fields.One2many(
+        'update.supplierprice.line',
+        'wizard_id',
+        string='Wizard lines')
+
+    @api.multi
+    def update_product_supplierprice(self):
+        invoice = self.env['account.invoice'].browse(
```
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/wizard/update_supplier_price.py]: self.ensure_one() is required if you do not manage a for each wizard in self.

File: account_invoice_product_supplier_price_update/wizard/update_supplier_price.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+            self._context['active_id'])
+        invoice.signal_workflow('invoice_open')
+        for line in self.wizard_line_ids:
+            if not line.suppinfo_id:
+                vals = {
```
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/wizard/update_supplier_price.py]: maybe calling default value function could allow to have different user defined default value ? (not sure about that point, just a question).

File: account_invoice_product_supplier_price_update/__openerp__.py [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    'version': '8.0.1.0.0',
+    'category': 'Accounting & Finance',
+    'website': 'http://akretion.com',
+    'author': 'Akretion, Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'installable': True,
+    'depends': [
+        'account',
+        'product_supplierinfo_tree_price_info',
+        'product_variant_supplierinfo'
```
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/__openerp__.py]: Hi. I understand why you depend on this module, but for me, the feature should work if product_variant_supplierinfo is not installed.
What about to split the module in two : 
-  account_invoice_product_supplier_price_update that will depends on account only and work with generic feature
- account_invoice_variant_product_supplier_price_update that will depends on 'product_variant_supplierinfo' and 'account_invoice_product_supplier_price_update', managing variants specifics, overloading function.

If people are agree with that point, I can help you for that refactoring.

File: account_invoice_product_supplier_price_update/__openerp__.py [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+               'the supplier price',
+    'version': '8.0.1.0.0',
+    'category': 'Accounting & Finance',
+    'website': 'http://akretion.com',
+    'author': 'Akretion, Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'installable': True,
+    'depends': [
+        'account',
+        'product_supplierinfo_tree_price_info',
```
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/__openerp__.py]: I don't see why you depend on this module. (even if this module is cool ;-) )

File: account_invoice_product_supplier_price_update/__openerp__.py [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+        'product_variant_supplierinfo'
+    ],
+    'data': [
+        'account_invoice_view.xml',
+        'wizard/update_supplier_price.xml'
+    ],
+    'demo': [
+        'demo/product_demo.xml',
+        'demo/account_invoice_demo.xml',
```
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/__openerp__.py]: could you git mv for
- demo/product_supplierinfo.xml
- demo/account_invoice.xml

[Ref](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#file-naming)

File: account_invoice_product_supplier_price_update/account_invoice.py [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+    _inherit = 'account.invoice'
+
+    @api.model
+    def _check_product_supplierprice(self):
+        lines = []
+        for line in self.invoice_line:
+            current_price_unit = []
+            suppinfo = self.env['product.supplierinfo'].browse(False)
+            supplier = self.partner_id.commercial_partner_id or self.partner_id
+            for seller in line.product_id.seller_ids:
```
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/account_invoice.py]: something like that should more rapid:

```
for seller in line.product_id.seller_ids.filtered(lambda seller: seller.name== supplier):
```

File: account_invoice_product_supplier_price_update/account_invoice.py [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+                if supplier == seller.name:
+                    suppinfo = seller
+                    current_price_unit = [pricelist.price for pricelist in
```
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/account_invoice.py]: Not a clear idea here, but I think we should think this algorithm as overloadable.

I think to [product_supplierinfo_discount](https://github.com/OCA/purchase-workflow/tree/10.0/product_supplierinfo_discount)

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Hi, @chafique-delli. Some remarks inline.  2 others below. Just a code review for the time being, I'll test it functionaly and more deeply friday.

Otherwise, thanks a lot to share this work !
For shops it is a "must have feature".
- Not sure the image module will help people to understand what the module makes. Default OCA module should be better ? (not a blocking point for me).
- Not sure to have understand all the workflow stuff : invoice validation... (but I just do a quick code review, I'll take a time friday).

**Comment by legalsylvain** [comment]
Hi @chafique-delli,
I took a deeply look on your PR. Some remarks :
* This module should be renamed better in account_invoice_supplierinfo_update because it's not only the price that is updated. (min_qty should be too)
* I still think that it should not depend on variant settings because it adds an unncessary dependency. It should be managed by glue module. The module you're sharing is very generic and is a must have feature, and supplierinfo for variant is specific. The current design will avoid some guys to install this module, or to have to disable manually some undesired features.
* This module doesn't work in all situation. Specially if there are some supplierinfo defined. (min_qty is not managed correctly).

I'm working on a PR on your PR, refactoring your PR. 

Thanks.

**Comment by legalsylvain** [comment]
hi all, 

FYI

If you're interested by this module, a pending refactoring is available here. 
https://github.com/akretion/account-invoicing/pull/3

**Review by legalsylvain — APPROVED** [review]

File: oca_dependencies.txt [hunk 23]
```diff
@@ -2,3 +2,5 @@ account-financial-tools
 stock-logistics-workflow
 product-attribute
 bank-payment
+product-variant
```
Comment (moylop260) [inline oca_dependencies.txt]: Where are required these repositories?

**Comment by legalsylvain** [comment]
Hi. 
I just rebased this PR, after the last merge. 
@chafique-delli, could you squach your commits ? (I can make it if you want).
@sebastienbeau, @all could you review this PR updated ?

Thanks !

**Comment by chafique-delli** [comment]
@legalsylvain , I want you to do it.Thank you.

File: oca_dependencies.txt (hunk 23 above)
Comment (legalsylvain) [reply oca_dependencies.txt]: Thanks for pointing this unused dependencies. I just fixed it.

**Comment by legalsylvain** [comment]
Note : Travis is red because upstream 8.0 branch is red.

**Review by sebastienbeau — APPROVED** [review]

**Comment by sebastienbeau** [comment]
Great work @legalsylvain @chafique-delli  ! Thx

**Comment by legalsylvain** [comment]
This PR has been strongly refactored and is now validated by 2 people. ( @sebastienbeau and me), but as I wrote a lot of the refactored code, I'd prefer to have a full external :+1: 

@eLBati , @pedrobaeza , @elicoidal , @moylop260 : is someone interested to finish to review that PR that you began to review ?

kind regards.

File: account_invoice_supplierinfo_update/models/account_invoice_line.py [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+        return supplierinfos and supplierinfos[0] or False
+
+    @api.multi
+    def _get_partnerinfo(self, supplierinfo):
+        """Given an invoice line and a supplierinfo, return the partnerinfo
+        that matches with line info, if exist"""
+        self.ensure_one()
+        if not len(supplierinfo.pricelist_ids):
+            return False
+        else:
```
Comment (moylop260) [inline account_invoice_supplierinfo_update/models/account_invoice_line.py]: If you have a `if-return` then the `else` is not needed

File: account_invoice_supplierinfo_update/models/account_invoice_line.py [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+
+    @api.multi
+    def _get_supplierinfo(self):
+        """Given an invoice line, return the supplierinfo that matches
+        with product and supplier, if exist"""
+        self.ensure_one()
+        supplierinfos = (self.product_id.seller_ids.filtered(
```
Comment (moylop260) [inline account_invoice_supplierinfo_update/models/account_invoice_line.py]: Unnecessary parents `(`

File: account_invoice_supplierinfo_update/wizard/wizard_update_invoice_supplierinfo_line.py [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+    price_variation = fields.Float(
+        string='Price Variation (%)', compute='_compute_price_variation',
+        digits_compute=dp.get_precision('Discount'))
+
+    state = fields.Selection(selection=_SELECTION_STATE)
+
+    @api.depends('current_price', 'new_price')
+    @api.multi
+    def _compute_price_variation(self):
+        for line in self:
```
Comment (moylop260) [inline account_invoice_supplierinfo_update/wizard/wizard_update_invoice_supplierinfo_line.py]: In order to reduce cyclomatic complexity (and quantity of tests to cover the code) you could remove `if-else` sentence using filtered and global write:
```python
self.write({'price_variation': False})
for line in self.filtered('current_price'):
     line.price_variation = 100 * (line.new_price - line.current_price) / line.current_price
```

File: account_invoice_supplierinfo_update_discount/models/account_invoice_line.py [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+    def _is_correct_partner_info(self, partnerinfo):
+        res = super(AccountInvoiceLine, self)._is_correct_partner_info(
+            partnerinfo)
+        return res and (self.discount == partnerinfo.discount)
+
+    @api.multi
+    def _prepare_supplier_wizard_line(self, supplierinfo, partnerinfo):
+        res = super(AccountInvoiceLine, self)._prepare_supplier_wizard_line(
+            supplierinfo, partnerinfo)
+        res['current_discount'] = partnerinfo and partnerinfo.discount or False
```
Comment (moylop260) [inline account_invoice_supplierinfo_update_discount/models/account_invoice_line.py]: new api supports BrowseNull.value <- `False`
Use: `res['current_discount'] = partnerinfo.discount`

File: account_invoice_supplierinfo_update_variant/wizard/wizard_update_invoice_supplierinfo_line.py [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+    to_variant = fields.Boolean(
+        string='Update to the variant',
+        help='- Check the option for an update on the product variant.\n'
+             '- Uncheck the option for an update on the product model.')
+
+    @api.multi
+    def _prepare_supplierinfo(self):
+        if self.to_variant:
+            return {
+                'product_id': self.product_id.id,
```
Comment (moylop260) [inline account_invoice_supplierinfo_update_variant/wizard/wizard_update_invoice_supplierinfo_line.py]: https://github.com/OCA/account-invoicing/blob/d0987a68bfc9c86a59e811c6262327d71cbafcb1/account_invoice_supplierinfo_update/wizard/wizard_update_invoice_supplierinfo_line.py#L62-L69

The original method have the same values but changing just `product_id`:
```python
    @api.multi
    def _prepare_supplierinfo(self):
        self.ensure_one()
        return {
            'product_tmpl_id': self.product_id.product_tmpl_id.id,
            'name': self.wizard_id.invoice_id.supplier_partner_id.id,
            'min_qty': 0.0,
            'delay': 1,
        }
```

What about use result from `super` and change to custom values required in pro of easiest inheriting?

`res = supers(...); res.update({...}); res.pop()`

(I don't know maybe is not possible)

**Review by moylop260 — APPROVED** [review]
👍 
Thanks!

Just small comments non-blocking

File: account_invoice_supplierinfo_update_discount/models/account_invoice_line.py (hunk 27 above)
Comment (legalsylvain) [reply account_invoice_supplierinfo_update_discount/models/account_invoice_line.py]: Your proposal doesn't work.

bool_error

But the following line does work.

        res['current_discount'] = partnerinfo and partnerinfo.discount

**Comment by legalsylvain** [comment]
@moylop260 : @chafique-delli and me updated our PR, according your last remarks.
Is it now all ok for you ? 

kind regards.

**Comment by legalsylvain** [comment]
Ping @moylop260 !

(And happy New year !)

**Comment by legalsylvain** [comment]
Previous review was OK. 3 :+1:. merging.

**Pull request merged**